components/eppp_link/eppp_uart.c

```diff
@@ -0,0 +1,213 @@
+#define PACKET_SIZE (MAX_PAYLOAD + 4)
+
+struct header {
+    uint8_t magic;
+    uint8_t check;
+    uint16_t size;
+} __attribute__((packed));
+
+static esp_err_t transmit(void *h, void *buffer, size_t len)
+{
+    struct eppp_handle *common = h;
+    struct eppp_uart *handle = __containerof(common, struct eppp_uart, parent);
+#ifndef CONFIG_EPPP_LINK_USES_PPP
+    static uint8_t out_buf[PACKET_SIZE] = {};
+    struct header *head = (void *)out_buf;
+    head->magic = HEADER_MAGIC;
+    head->check = 0;
+    head->size = len;
+    head->check = HEADER_MAGIC + len + (len >> 8);
+    memcpy(out_buf + sizeof(struct header), buffer, len);
+    ESP_LOG_BUFFER_HEXDUMP("ppp_uart_send", out_buf, len + sizeof(struct header), ESP_LOG_DEBUG);
+    uart_write_bytes(handle->uart_port, out_buf, len + sizeof(struct header));
+#else
+    ESP_LOG_BUFFER_HEXDUMP("ppp_uart_send", buffer, len, ESP_LOG_DEBUG);
+    uart_write_bytes(handle->uart_port, buffer, len);
+#endif
+    return ESP_OK;
+}
+
+static esp_err_t init_uart(struct eppp_uart *h, struct eppp_config_uart_s *config)
+{
+    h->uart_port = config->port;
+    uart_config_t uart_config = {};
+    uart_config.baud_rate = config->baud;
+    uart_config.data_bits = UART_DATA_8_BITS;
+    uart_config.parity    = UART_PARITY_DISABLE;
+    uart_config.stop_bits = UART_STOP_BITS_1;
+    uart_config.flow_ctrl = UART_HW_FLOWCTRL_DISABLE;
+    uart_config.source_clk = UART_SCLK_DEFAULT;
+
+    ESP_RETURN_ON_ERROR(uart_driver_install(h->uart_port, config->rx_buffer_size, 0, config->queue_size, &h->uart_event_queue, 0), TAG, "Failed to install UART");
+    ESP_RETURN_ON_ERROR(uart_param_config(h->uart_port, &uart_config), TAG, "Failed to set params");
+    ESP_RETURN_ON_ERROR(uart_set_pin(h->uart_port, config->tx_io, config->rx_io, UART_PIN_NO_CHANGE, UART_PIN_NO_CHANGE), TAG, "Failed to set UART pins");
+    ESP_RETURN_ON_ERROR(uart_set_rx_timeout(h->uart_port, 1), TAG, "Failed to set UART Rx timeout");
+    return ESP_OK;
+}
+
+static void deinit_uart(struct eppp_uart *h)
+{
+    uart_driver_delete(h->uart_port);
+}
+
+#ifndef CONFIG_EPPP_LINK_USES_PPP
+static void process_packet(esp_netif_t *netif, uint8_t *buffer, size_t len)
+{
+    static uint8_t in_buf[PACKET_SIZE + UART_BUF_SIZE] = {};
+    static size_t current_offset = 0;
+    struct header *head = (void *)in_buf;
+
+    // Add received data to our buffer
+    if (current_offset + len <= sizeof(in_buf)) {
+        memcpy(in_buf + current_offset, buffer, len);
+        current_offset += len;
+    } else {
+        ESP_LOGW(TAG, "Buffer overflow, discarding data");
+        current_offset = 0;
+        return;
+    }
+
+    // Process while we have enough data for at least a header
+    while (current_offset >= sizeof(struct header)) {
+        // Verify magic number
+        if (head->magic != HEADER_MAGIC) {
+            // Try to recover by shifting one byte at a time
+            memmove(in_buf, in_buf + 1, current_offset - 1);
```

Comment:
It seems like inefficient way of how to find a new header's start, in my opinion. Too much copying involved. From `memmove` description:  "The objects may overlap: copying takes place as if the characters were copied to a temporary character array and then the characters were copied from the array to dest."

Reply:
reworked